### ISSUE TITLE
Question about interactions between client and server and vue componenets executing JS code

### ISSUE THREAD

**Issue by JovanVeljanoski** [widgetti/solara#838]
Hi, 

This might be a strange question/request but please bear with me..

I've found out that for some cases, having something executed by the front-end is "better" than on the server side. An example that comes to mind is, say your application needs to play a sounds. If you are gonna deploy this, it seems to me easier to write some JS code to play sound using the browser capabilities, than fight with device ports etc.. one is gonna do the sounds generation on the server side.

Solara allows this actually, with the `@component_vue` one can make a, say button that when pressed plays a sounds. An example of this is (kind of pseudo code)

```python
@solara.component_vue("play_audio.vue")
def PlayAudio(audio_data: list[bytes], label: str = 'Listen', style: str = ''):    ...
```

And the corresponding vue component will be something like

```vue
<template>
  <div>
    <audio ref="audioPlayer" :src="audioSrc" style="display: none;"></audio>
    <v-btn @click="playAudio" :disabled="!audioSrc" :color="color" :style="style">
      <v-icon left>mdi-volume-high</v-icon>
      {{ label }}
    </v-btn>
  </div>
</template>

<script>
export default {
  props: {
    audio_data: {
      type: Array,
      required: true
    },
    label: {
      type: String,
      default: 'Listen'
      },
    color: {
      type: String,
      default: 'primary'
    },
    style: {
      type: String,
      default: ''
    },
  },
  
  data() {
    return {
      audioSrc: null,
    };
  },
  
  playAudio() {
   // implementation of this
    },
    
 // whatever else is needed for the JS to work
</script>
```

This works fairly well! 

My problem is the following:
 - It is kind of once-directional. If I make a component in this way, it is can not communicate with the rest of the solara (python) code. For instance, say I want to play some sound with the approach above, but at the same time I want to.. for example disable some button that is written in `solara.Button()` way (so in python). 
 - Of course I can make the entire "UI" in vue, and then I can disable another button when the play button is played. But that is just writing fronted code (from my basically 0 knowledge of that), and is kind of moving away from the power/elegance of solara.
 - Is it possible to write some kind of JS function.. that can be executed from `solara.Button(on_click)` instead? Maybe somehow wrapped in python. That will allow for making solara.task that I find useful for things. And one can "wrap" it so that, some things can be executed on the server, and relevant stuff can be dispatched to the browser.
 - Ultimately (pretty sure this is not possible.. and maybe completely along the wrong lines of thinking.. but sometimes I think it would be awesome if.. the `@component_vue` could be made 2-directional with the python server.. i.e... you dispatch something to the frontend (clinet) a computation is done there, or some process or whatever, and the result can be sent back to the server for something else to be updated (like a state, or whatever needs to happen next). 
 
 I hope at least some of these things make sense.. 
 
 looking forward to some kind of answer / advice / or pointing in the right direction. Thank you!

**Comment by maartenbreddels** [comment]
Hi Jovan,

another good question.
I am working on documentation that elaborates a bit more on that. But given that our docs at https://solara.dev/documentation/api/utilities/component_vue didn't have examples, I expanded this a bit at #844

However, I think that most of the questions are answered by #847 especially by the last example on the vue component howto.

**Comment by JovanVeljanoski** [comment]
Thanks!  That's very helpful. I am still not sure how the changes are propagated back to the server from the vue component though, i don't think there is an example in the documentation for that. 

But I think I did find an example from you somewhere  ( i think it was the example of the local_storage access), and maybe that has a hint.. will analyze it more in depth when I have a chance! 

Thank you!

**Comment by JovanVeljanoski** [comment]
Hi again, 

Since I could not find an example of the following.. can maybe someone try to explain what I am doing wrong (or if what I am trying to do is possible). 

[Link to pycafe example](https://py.cafe/snippet/solara/v1?pycafe-app-view=false&pycafe-edit-enable=true#c=H4sIAFSQNGcEA6VXbW_juBH-K4IOhbOorViy82ZEh8UtWrQfrijaQ_uhLgJaom1uJFJLUkl8Qf77PTOUZDnJYvdw3EVszgyfeeG80M9xYUoZr2JVN8b6yJlKWLHWa_0xfE0KA46W2t89tPJskpzj487XTYLPyYe1LuU2-vnwqZc6exBVK1eR8zbKo8lkGlViI6uB8BftpY28fPJgGX3XiReiglwlIfIP4EyjUm7a3SraGFOB9ldROflhtdYRVpIk7xoYbPmn2MmzXpTRcb4TbZ28s1IUXj3AFbKehHRbf0VkDokgQ8h7octKBovvCuygSMvHQOg10mqsQiC26_g_rD6IlpE3q-h5OPCyjjv9tJiUOOlHiL3uR-X3vXmfTNXWevCP1glb2PLM-UMl8-dJLZ5mj6r0-8kqmlzP583T5GV8kNbJ4bfY_Rpf8FsuLTY5D27w3-n7chVlQ76OQx44U0tOhnX8Ffk-RfJ3ov-VI5w6-S_2XRtGMe9X5_7Pwt6X5lGfbSe_7GWXOsrhzvhr8OqlT5rx6gB-ar03-ix4OPm7LqysEbCQ50Wlivu8EvWmFCtKuXDZ-AyK_hylH6YoErq6dczXtoqyiz-dZEm_3rcYYBvElE0egNngeBpb-aVVwSCHcg8IzPGHhhpAoGDfHPzeaFCagylVKWcP8yS7SFKwKnEwrY9Xz_GDtE6RVAZoY_y_zCOROyyL3TQu9qoqrYTQ_waOFxv4DSZ7GK_Si_k0rpX-b9guwu5vUu320ENbVeLYVlXyJ6A6aT8Z7YXS0n5FA4nONkEWIo0g3Dh--f_LdJAZrHgPAMzjOdE0SXOIT88e2X0vJJGxgu8IwLed7pz4lsO9rycGDGr2HWp6dfE9OlGUIIvqW0p7OdJK_16mHHikFmS0qEmmDw4-CQWISD7Qb72sm0p4-SNl9m2pHvgLvj7MqBfMtkpWZSChscxqTChqGZXBlOCWuo577kelm9aD2TYlEF9xV1XXbfizp_94ez5WdFS-8Tr6GAp1HYd-E2r6E9HWcVSYyljw0OBrYQ_ruDsbRf-WPgrKAxg0AI3Zt-fBw9vzkd_YusKqxuM7_GsrmcgnmsAOo-g5gMAhcfah39Gy0rdWjym0jnFZRX6vXMKFf5R5CV9fum5YS9R3SU3iKDKK3qlGWmp7xrDcV99waeFu0T5kUpnd2WSEtULv46NHE191s842Wr19tN4E_w9b9QZxGjr87zKRVq-3H0zvqj6RCPLfg9997SPxKHyBKTDCD3h4IgSgMYvWK0VIpV70KNfrWGv8vz0fshCVjAZvinuqYyr_UKw_4C4cjCjnl9dzmd2km2WWbsXmclOmi-t0vlxeZ9cXV9u1FvqgTJ4vk2Uyx87ujM5mxXar8jxbJOkr4myjdKn0zoGbJlkyj36IamgnITS0PE-TBR9xuLF7qUkOyCko3lvadZgbUdzD4SrPMaaYIEXr1batnGmbJRmUZkkGeoVn3T7PL5M5iRXYSI_nJbAuoAvIhbQ4SPbOgX6VZDgVHEiTFBMQu72w6KszbWwtKvWrtHm-wGESpKTK82tYdYWdqetgEVicos2BYMAeHC1lYazw6CgwABqIssX7s3yq2ZsrIsknWcAdvSP_2eEtQvLZIbdhfy2InLFDdG90f2QRrF2C8qXUpJRN36cpgaZ8O_u2Fhrmc3T4tCo1sBZ0TDWHe2k1GidihckPD0CiR0FO7l0SWkeY7aSGdRTDLvpgoN9zxNilsEdY2f1OV3PA-N9Jj2MUrwuQHJoggoE3BYWfDkPwsywRfOxueKf0Z5Gxe2QTxaAxeGmHS2DkcVzg2VWyAK2FrdLOcEFIadIIr0ZkY2HuBV33kSgfIMpOXY4hMGbx8OGbgLrhIk_ZGC2WZyOfR2oNAphCs-aw4zcY8TjBR7whKORNinMYMvzAIwqSqt_OlAfO4DSR28aJLfygHOGDvqmMr9RmpnSFOU7qODFrhJmzi2NYK-db4hIU9npTVMI5VfBljVwkRggfTrJWUIyGuyAhUS8pcfRmS3UBCjJuzhTp_Ey4gy6oNaSIJp00XuJ3HRL1EuEZ6ejoM7dXXfEQRFuH0slYh4FKi1cponSFOwNcg85uCmQ_oghql-6NsA4q56g4xKPBcJUFDFsmSKVBY4OSxaMFRc0BuqJMbFRVcf-BlwyPiU1ulYrQlzAKcWosfrj4vWzdkFYwN8iD1fgZdZZ7BTpFdonAN44KhUJzk1wfLcB7yppCOmCTAQTQWokEFHxLGVt_KMgdyZmXkXNDEuG-uR6bQ82ZgVcNeiWqCDx4cEOc8Irfe98ERBbH891RiycSyou1cEXT8A6WZogd-RroVFozjNJdMAPGEusgqFdRS2XUX-svYKLFjILMv6YNcjgNd0G_RJAWbPyCy9xui8VicTODzwrqDRQAj26bODfXl6850GVVgV6eLoBJCNjRTVD369U0peMqwQHIDOY4qcvMWwy1QRQdvW26UXBJDQleO_VEfEQX8E4rjAHOYB4UeFbgtxJvuTOF_dAeXpFbHiA9CZWq0F7Ie3QoUDwGGEUdeF27AclW0ntOygW3VeQ29ZSSUxr-4CCGwqFwDg0R4CxjbJDgQoGLygMFPqEgloRLb3eKyZt7hmnI3mw5v6bsghjmzUkqLZGa4LRWobuFR-ygFv0EfWZBOHyb7YNCT8V9w3Y4tNb8jCkNJthpYjCdphYUQD8OH1kF_z7s8hWmP0q0Gzy-IYr04jsHSWLOdo8I6rSwEEQHBIzovi7pjoN0YEAWWYMgghb6LV4MvavBRFLoXIU_XyrlJXbxy2_-kM6iKRMAAA)

Basically, i have a vue component, and when I click a button I want some output to be synced to a solara reactive component. 

In this case, i would like a simple thing.. the component is a textbox and a button, so when the button is clicked, the value of the input field to be synced to a reactive variable.

If this is possible, would be nice to have an explicit example on this. My apologies if this is covered in an example somewhere that i have missed

**Comment by iisakkirotko** [comment]
Hey!

The issue you're running into has to do with naming of the arguments - argument names like `on_foo` are automatically tied to the corresponding argument `foo`, and will be called whenever the value of `foo` changes on the javascript side. However, you're not actually changing the value of `value`, which means that `on_value` never gets called. Note that `on_foo` from python doesn't actually get mapped to functions `on_foo` in javascript. 

The easiest fix is therefore to change from calling `this.on_value` in the vue template to simply assigning to `this.value`, which will trigger `on_value` in python, like I've done [in this py.cafe project](https://py.cafe/iisakkirotko/solara-text-input-demo).

Another option would be to make use of one of the other type of magic argument, `event_bar`, which will get mapped to a function `bar` in javascript. You can find the same example using this method [here](https://py.cafe/iisakkirotko/solara-text-number-app).

**Comment by JovanVeljanoski** [comment]
Hi! 

Thank you @iisakkirotko ! This was exactly the explanation I was looking for! Thank you again! 
If this can somehow find its way into the docs, would be amazing! 

(feel free to close this issue when/if you see fit)
